pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+# Code intelligence
+
+We'd like to replace elixir_sense with our own implmentation for the following bits
+of code intelligence.
+
+ * Go to definition
+ * Find references
+ * Refactoring
+
+In order to have this level of intelligence of elixir code, we need to
+search both beam files and source code to discover relevant parts and
+locations of elixir and erlang code, index them an make them
+searchable. This is a large task that will not be accomplished in one
+commit, and this document will help to lay out the scope of the work,
+as well as collect thoughts and drive the implementation.
+
+Code intelligence will be oragnized into the following discrete areas:
+
+* **Discovery** - Finding points of interest in the project's codebase, dependencies and elixir core modules
+* **Caching** - Writing the discovered items to disk and ensuring the index is up to date with the state of the project at *all times*. The goal here is to never have to delete the `.lexical` directory.
+* **Indexing** - Storing the discovered artifacts in a local in-memory index that enable fast searching.
+* **Searching** ideally using something like the [sublime text's fuzzy search](https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/), but we definitely need some type of fuzzy search.
```

Comment:
so IME the LS does not have to worry about the mechanism for search, LS instead worries about answering queries defined by the LSP. It is up to the client _how_ to search through the answers from the server, e.g. fuzzy search through the symbols in a document, or grep through them.

Are you thinking of some specific `search` mechanism _not_ defined in the LSP?

Reply:
I think it would also be useful to have other types of searches available. Given a line of code, to which function/module does it belong? Which modules use this macro etc.

Reply:
>  It is up to the client how to search through the answers from the server, e.g. fuzzy search through the symbols in a document,

Well, sorta. If the language server doesn't offer any suggestions, then the client can't filter them further. Consider what would happen if we did a strict equality and the client searched for `ENum`, and we returned `[]` for completions.

Reply:
> I think it would also be useful to have other types of searches available.

Yes, definitely, consider the above a starting point.

Reply:
Cool! Perhaps it would be useful to start enumerating in a doc the queries that we support and mark them as whether they are queries defined by the LSP or something more custom